src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp

```diff
@@ -0,0 +1,730 @@
+    {
+        ClassDesc2* class_descr(new OSLPluginClassDesc(&shader));
+        ParamBlockDesc2* param_block_descr(new ParamBlockDesc2(
+            // --- Required arguments ---
+            0,                                          // parameter block's ID
+            L"oslTextureMapParams",                     // internal parameter block's name
+            0,                                          // ID of the localized name string
+            class_descr,                                // class descriptor
+            P_AUTO_CONSTRUCT,                           // block flags
+
+                                                        // --- P_AUTO_CONSTRUCT arguments ---
+            0,                                          // parameter block's reference number
+            p_end
+        ));
+
+        int ctrl_id = 100;
+        int param_id = 0;
+        for (auto& param_info : shader.m_params)
+        {
+            add_parameter(
+                param_block_descr,
+                param_info,
+                param_info.m_max_param,
+                shader.m_string_map,
+                shader.m_texture_id_map,
+                shader.m_material_id_map,
+                param_id,
+                ctrl_id);
+        }
+
+        shader.m_output_param = add_output_parameter(
+            param_block_descr,
+            shader.m_output_params,
+            shader.m_string_map,
+            param_id,
+            ctrl_id);
+
+        auto tn_vec = shader.findParam("Tn");
+
+        if (!shader.m_is_texture && tn_vec != nullptr)
+            add_bump_parameters(
+                param_block_descr,
+                shader.m_texture_id_map,
+                param_id);
+
+        m_paramblock_descriptors.push_back(MaxSDK::AutoPtr<ParamBlockDesc2>(param_block_descr));
+        m_class_descriptors.push_back(MaxSDK::AutoPtr<ClassDesc2>(class_descr));
+    }
+}
+
+void OSLShaderRegistry::add_parameter(
+    ParamBlockDesc2*        pb_desc,
+    const OSLParamInfo&     osl_param,
+    MaxParam&               max_param,
+    IdNameMap&            string_map,
+    IdNameVector&            texture_map,
+    IdNameVector&            material_map,
+    int&                    param_id,
+    int&                    ctrl_id)
+{
+    const int string_id = ctrl_id++;
+    string_map.insert(std::make_pair(string_id, utf8_to_wide(max_param.max_label_str)));
+
+    max_param.max_ctrl_id = string_id;
+    max_param.max_param_id = param_id;
+
+    auto param_str = utf8_to_wide(max_param.osl_param_name);
+    if (max_param.param_type == MaxParam::Color)
+    {
+        Color def_val;
+        if (osl_param.hasDefault)
+        {
+            const float r = static_cast<float>(osl_param.defaultValue.at(0));
+            const float g = static_cast<float>(osl_param.defaultValue.at(0));
+            const float b = static_cast<float>(osl_param.defaultValue.at(0));
+            def_val = Color(r, g, b);
+        }
+
+        int ctrl_id_1 = ctrl_id++;
```

Comment:
Move this into the `AddParam()` call, this will avoid the other variables named `ctrl_id_1` below from shadowing this one.

Reply:
I don't understand how I should move it into the call.
Those ctrl_id_1-s are in different if() branches. Should I add `else` in the end of each `if` to show that only one branch is executed at a time?

Reply:
Forget it, I was confused by the indentation. There is no shadowing.